ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Parser.hs

```diff
@@ -79,74 +71,81 @@ epochFileParser'
   -> EpochFileParser
        (EpochFileError hash)
        m
-       (Secondary.Entry hash, WithOrigin hash)
+       (Secondary.Entry hash)
+       hash
 epochFileParser' getSlotNo getHash getPrevHash hasFS decodeBlock isEBB
                  getBinaryInfo isNotCorrupt =
-    EpochFileParser $ \fsPath k ->
+    EpochFileParser $ \fsPath expectedChecksums k ->
       Util.CBOR.withStreamIncrementalOffsets hasFS decoder fsPath
-        (k . checkIfHashesLineUp . S.map extractEntry . stopAtCorruption)
+        (k . checkIfHashesLineUp . checkEntries expectedChecksums)
   where
-    decoder :: forall s. Decoder s (BL.ByteString -> BinaryInfo (BlockInfo hash))
-    decoder = extractBlockInfo <$> decodeBlock
-
-    -- | It is important that we don't first parse all blocks, storing them
-    -- all in memory, and only /then/ extract the information we need. So
-    -- make sure we don't create thunks refering to the whole block.
-    extractBlockInfo :: (BL.ByteString -> blk)
-                     -> (BL.ByteString -> BinaryInfo (BlockInfo hash))
-    extractBlockInfo f bs =
-        getBinaryInfo blk $> BlockInfo
-          { hash         = getHash blk
-          , prevHash     = getPrevHash blk
-          , checksum     = computeCRC bs
-          , blockOrEBB   = case isEBB blk of
-            Just epoch -> EBB epoch
-            Nothing    -> Block (getSlotNo blk)
-          -- TODO do this only when the checksum doesn't match the known one
-          , notCorrupted = isNotCorrupt blk
-           }
-      where
-        blk = f bs
-
-    stopAtCorruption
-      :: Stream (Of (Word64, (Word64, BinaryInfo (BlockInfo hash))))
+    decoder :: forall s. Decoder s (BL.ByteString -> (blk, CRC))
+    decoder = decodeBlock <&> \mkBlk bs ->
+      let !blk      = mkBlk bs
+          !checksum = computeCRC bs
+      in (blk, checksum)
+
+    -- | Go over the expected checksums and blocks in parallel. Stop with an
+    -- error when a block is corrupt. Yield correct entries along the way.
+    --
+    -- If there's an expected checksum and it matches the block's checksum,
+    -- then the block is correct. Continue with the next.
+    --
+    -- If they do not match or if there's no expected checksum in the stream,
+    -- check the integrity of the block (expensive). When corrupt, stop
+    -- parsing blocks and return an error that the block is corrupt. When not
+    -- corrupt, continue with the next.
+    checkEntries
+      :: Stream (Of  CRC) m ()
```

Comment:
I'd prefer this to be a list just to facilitate a potential future changeover to `conduit`.